std/functional.d

```diff
@@ -899,6 +903,152 @@ template partial(alias fun, alias arg)
     assert(result2 == 1.5f);
 }
 
+/**
+Takes a function of (potentially) many arguments, and returns a function taking
+one argument and returns a callable taking the rest.  f(x, y) == curry(f)(x)(y)
+
+Params:
+    F = a function taking at least one argument
+    t = a callable object whose opCall takes at least 1 object
+Returns:
+    A single parameter callable object
+*/
+template curry(alias F)
+{
+    //inspired from the implementation from Artur Skawina here:
+    //https://forum.dlang.org/post/mailman.1626.1340110492.24740.digitalmars-d@puremagic.com
+    //This implementation stores a copy of all filled in arguments with each curried result
+    //this way, the curried functions are independent and don't share any references
+    //eg: auto fc = curry!f;  auto fc1 = fc(1); auto fc2 = fc(2); fc1(3) != fc2(3)
+    import std.traits : isCallable;
+    static if (isCallable!F)
+    {
+        auto curry()
+        {
+            import std.traits : Parameters;
+            alias FParams = Parameters!F;
+
+            struct CurryImpl(size_t N)
```

Comment:
Could reside outside of `auto curry()...` or at least be static to avoid the uneccessary context pointer